flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/GroupAggProcessFunctionWithUpdateInterval.scala

```diff
@@ -0,0 +1,204 @@
+  protected var state: ValueState[Row] = _
+
+  // counts the number of added and retracted input records
+  protected var cntState: ValueState[JLong] = _
+
+  // stores the input for group keys
+  private var inputState: ValueState[Row] = _
+
+  // stores the last emit row
+  private var emitState: ValueState[Row] = _
+
+  // stores the emit time
+  private var emitTimerState: ValueState[JLong] = _
+
+
+  override def open(config: Configuration) {
+    LOG.debug(s"Compiling AggregateHelper: ${genAggregations.name}\n\n" +
+      s"Code:\n${genAggregations.code}")
+    val clazz = compile(
+      getRuntimeContext.getUserCodeClassLoader,
+      genAggregations.name,
+      genAggregations.code)
+    LOG.debug("Instantiating AggregateHelper.")
+    function = clazz.newInstance()
+
+    newRow = new CRow(function.createOutputRow, true)
+    prevRow = new CRow(function.createOutputRow, false)
+    typeSerializer = outputRowType.createSerializer(new ExecutionConfig())
+
+    state = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregateState", aggregationStateType))
+    cntState = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong]("GroupAggregateInputCounter", Types.LONG))
+    inputState = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregatePreEmitState", classOf[Row]))
+    emitState = getRuntimeContext.getState(
+      new ValueStateDescriptor[Row]("GroupAggregateEmitState", outputRowType))
+    emitTimerState = getRuntimeContext.getState(
+      new ValueStateDescriptor[JLong]("emitTimeState", Types.LONG))
+
+    initCleanupTimeState("GroupAggregateWithUpdateIntervalCleanupTime")
+  }
+
+  override def processElement(
+      inputC: CRow,
+      ctx: ProcessFunction[CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    val currentTime = ctx.timerService().currentProcessingTime()
+    // register state-cleanup timer
+    registerProcessingCleanupTimer(ctx, currentTime)
+
+    val input = inputC.row
+
+    // get accumulators and input counter
+    var accumulators = state.value
+    var inputCnt = cntState.value
+
+    if (null == accumulators) {
+      accumulators = function.createAccumulators()
+      inputState.update(input)
+      inputCnt = 0L
+    }
+
+    // update aggregate result and set to the newRow
+    if (inputC.change) {
+      inputCnt += 1
+      // accumulate input
+      function.accumulate(accumulators, input)
+    } else {
+      inputCnt -= 1
+      // retract input
+      function.retract(accumulators, input)
+    }
+
+    state.update(accumulators)
+    cntState.update(inputCnt)
+
+    var triggerTimer = emitTimerState.value
+
+    if (null == triggerTimer) {
+      triggerTimer = 0L
+    }
+
+    if (currentTime >= triggerTimer) {
+
+      val newTimer = currentTime + queryConfig.getUnboundedAggregateUpdateInterval
+
+      emitTimerState.update(newTimer)
+
+      ctx.timerService().registerProcessingTimeTimer(newTimer)
```

Comment:
I think you are right. For now, a processing-time timer is sufficient and maybe we don't need an event-time timer at all. This parameter does not affect the semantics of the result, so processing time is fine and easier to reason about. I also agree that a single configuration parameter is enough. If `updateInterval` is not set or set to `0`, results are immediately updated. Otherwise, according to the specified interval.

However, I still think that we need three different implementations of the operator:
1. Immediate update (the one that we already have)
2. Interval update for processing time, i.e., no watermarks that need to be hold back
3. Interval update for event time, i.e., holding back watermarks. The updates should always have the watermark of the last update, which means we have to hold back the watermarks. This operator cannot be implemented with a `ProcessFunction` and requires a custom operator.

The cases 2. and 3. are determined based on the time characteristics setting of the `StreamExecutionEnvironment`.